libfreerdp/core/smartcardlogon.c

```diff
@@ -0,0 +1,404 @@
+#include <openssl/objects.h>
+
+#include <winpr/error.h>
+#include <winpr/ncrypt.h>
+#include <winpr/string.h>
+#include <winpr/wlog.h>
+#include <winpr/smartcard.h>
+#include <winpr/crypto.h>
+
+#include <freerdp/log.h>
+
+#include "smartcardlogon.h"
+
+#define TAG FREERDP_TAG("smartcardlogon")
+
+static BOOL getAtr(LPWSTR readerName, BYTE* atr, DWORD* atrLen)
+{
+	WCHAR atrName[256];
+	DWORD cbLength;
+	DWORD dwShareMode = SCARD_SHARE_SHARED;
+	DWORD dwPreferredProtocols = SCARD_PROTOCOL_T0 | SCARD_PROTOCOL_T1;
+	SCARDHANDLE hCardHandle;
+	DWORD dwActiveProtocol = 0;
+	BOOL ret = FALSE;
+	LONG status = 0;
+	SCARDCONTEXT scContext;
+
+	status = SCardEstablishContext(SCARD_SCOPE_USER, NULL, NULL, &scContext);
+	if (status != ERROR_SUCCESS || !scContext)
+		return FALSE;
+
+	status = SCardConnectW(scContext, readerName, dwShareMode, dwPreferredProtocols, &hCardHandle,
+	                       &dwActiveProtocol);
+	if (status != ERROR_SUCCESS)
+		goto out_connect;
+
+	*atrLen = 256;
+	status = SCardGetAttrib(hCardHandle, SCARD_ATTR_ATR_STRING, atr, atrLen);
+	if (status != ERROR_SUCCESS)
+		goto out_get_attrib;
+
+	cbLength = 256;
+	status = SCardListCardsW(scContext, atr, NULL, 0, atrName, &cbLength);
+	if (status != ERROR_SUCCESS)
+		goto out_listCards;
+
+	/* WLog_DBG(TAG, "ATR name: %ld -> %S\n", cbLength, atrName); */
+	ret = TRUE;
+out_listCards:
+out_get_attrib:
+	SCardDisconnect(scContext, SCARD_LEAVE_CARD);
+out_connect:
+	SCardReleaseContext(scContext);
+	return ret;
+}
+
+void smartcardCert_Free(SmartcardCert* scCert)
+{
+	free(scCert->reader);
+	crypto_cert_free(scCert->certificate);
+	free(scCert->pkinitArgs);
+	free(scCert->containerName);
+	free(scCert->upn);
+	free(scCert->userHint);
+	free(scCert->domainHint);
+	free(scCert->subject);
+	free(scCert->issuer);
+
+	ZeroMemory(scCert, sizeof(*scCert));
+}
+
+static BOOL treat_sc_cert(SmartcardCert* scCert)
+{
+	scCert->upn = crypto_cert_get_upn(scCert->certificate->px509);
+	if (scCert->upn)
+	{
+		size_t userLen;
+		const char* atPos = strchr(scCert->upn, '@');
+
+		if (!atPos)
+		{
+			WLog_ERR(TAG, "invalid UPN, for key %s (no @)", scCert->containerName);
+			return FALSE;
+		}
+
+		userLen = (atPos - scCert->upn);
+		scCert->userHint = malloc(userLen + 1);
+		scCert->domainHint = strdup(atPos + 1);
+
+		if (!scCert->userHint || !scCert->domainHint)
+		{
+			WLog_ERR(TAG, "error allocating userHint or domainHint, for key %s",
+			         scCert->containerName);
+			return FALSE;
+		}
+
+		memcpy(scCert->userHint, scCert->upn, userLen);
+		scCert->userHint[userLen] = 0;
+	}
+
+	scCert->subject = crypto_cert_subject(scCert->certificate->px509);
+	scCert->issuer = crypto_cert_issuer(scCert->certificate->px509);
+	return TRUE;
+}
+
+#ifndef _WIN32
+static BOOL build_pkinit_args(rdpSettings* settings, SmartcardCert* scCert)
+{
+	/* pkinit args only under windows
+	 * 		PKCS11:module_name=opensc-pkcs11.so
+	 */
+	size_t sz = strlen("PKCS11:module_name=:slotid=XXXXX");
```

Comment:
you can do this with just calling `snprintf` with a `NULL` buffer and `0` length (check the return)

Reply:
good idea, will do that.